### PR TITLE
[inductor] do not generate loops when the condition doesn't hold

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98185

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Codegen sometimes prints loops that would never execute because the loop condition is false. This PR removes them.

I haven't looked into why this happens in the first place (might be a problem somewhere else). At least this makes codegen output nicer to read.

Example (test currently xfail'd because there's another non-dynamic loop variable there):
```
pytest -rfxs test/inductor/test_torchinductor_codegen_dynamic_shapes.py  -k test_as_strided_dynamic_shapes
...
-        #pragma omp simd simdlen(4)
-        for(long i0=static_cast<long>(4096); i0<static_cast<long>(4096); i0+=static_cast<long>(1))
-        {
-            auto tmp0 = out_ptr0[static_cast<long>(i0)];
-            auto tmp1 = static_cast<float>(2);
-            auto tmp2 = tmp0 + tmp1;
-            out_ptr1[static_cast<long>(i0)] = tmp2;
-        }

```

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -2615,6 +2618,10 @@ def lines(self):
         offset_str = f"{INDEX_TYPE} {self.var}={cexpr_index(self.offset)}"
         size_str = f"{self.var}<{cexpr_index(self.size)}"
         steps_str = f"{self.var}+={cexpr_index(self.steps)}"
+        if cexpr_index(self.offset) == cexpr_index(self.size):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2621]: Does it work with dynamic shapes? Also, how about move this check to the beginning of the method?

**Comment by jgong5** [comment]
> I haven't looked into why this happens in the first place (might be a problem somewhere else). At least this makes codegen output nicer to read.

Yes, these loops can be ignored. They are not of a performance concern since compiler will optimize them out anyway. We didn't remove them due to debuggability and test coverage: 1) scalar kernel is always generated, which is easier to read than vectorized counterparts; 2) it can uncover kernel codegen issues - all the kernels are generated and compiled, even if they are not executed. I'm OK to remove them by default. But can you add an option to always generate them? E.g., some configuration under `config.cpp` and its default value controlled by some environment variable? So, we can turn it on if needed. cc @EikanWang

**Comment by ezyang** [comment]
Deferring to the cpu inductor crew

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (nkaretnikov) [reply torch/_inductor/codegen/cpp.py]: Done. Yes, this should work with everything. I'm using `==` on strings.

If I were to use this instead, it wouldn't work:

```
 if config.cpp.no_redundant_loops and self.offset >= self.size

...

TypeError: cannot determine truth value of Relational
```

**Comment by nkaretnikov** [comment]
@jgong5 Updated. LMK what you think.

**Review by jgong5 — APPROVED** [review]
LGTM.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**